WIP layout for ID card preview

rianfowler opened the pull request:
Tried to make all of the positioning percentage based. The background / mockup isn't exactly 3:2 but think I've got it mostly right.

Some questions:

- ~~I'm not sure where to store the static assets. I found the dev watch is being pulled from a build/development folder~~
- ~~I need a VA logo with a transparent background. I got an .eps from @stacyhlinn but I'm not sure how to export it correctly.~~
- There are some differences between design and implementation that may be hard to reconcile e.g. font weights, Myriad font. Will follow-up with designers.
- Other questions inline
<img width="584" alt="screen shot 2018-02-12 at 20 13 07" src="https://user-images.githubusercontent.com/4998130/36131539-e77dcb84-1037-11e8-91b0-af1840339370.png">
<img width="596" alt="screen shot 2018-02-12 at 20 12 27" src="https://user-images.githubusercontent.com/4998130/36131540-e78d9c08-1037-11e8-931b-df8bcf4737ab.png">
<img width="587" alt="screen shot 2018-02-12 at 20 14 40" src="https://user-images.githubusercontent.com/4998130/36131541-e79ec032-1037-11e8-95fa-614c1f756c1b.png">
<img width="584" alt="screen shot 2018-02-12 at 20 14 26" src="https://user-images.githubusercontent.com/4998130/36131542-e7b541ea-1037-11e8-8033-ed5c3e583c9d.png">
<img width="590" alt="screen shot 2018-02-12 at 20 14 08" src="https://user-images.githubusercontent.com/4998130/36131543-e7c92c5a-1037-11e8-9d3c-b4895f9da786.png">
<img width="587" alt="screen shot 2018-02-12 at 20 13 52" src="https://user-images.githubusercontent.com/4998130/36131544-e7e16a4a-1037-11e8-85dc-d1098743d33e.png">


Here's the original for comparison. 
![vic-2 3 cropped](https://user-images.githubusercontent.com/4998130/36008121-5e95a42e-0d0b-11e8-90fa-46675563e6d9.png)

Commit:
layout for ID card preview

rianfowler commented on src/js/vic-v2/components/VeteranIDCard.jsx:
```diff
@@ -1,7 +1,34 @@
+      <svg width={'100%'} height={'100%'}>
+        <image width={'100%'} href={'/img/vic-preview-background.svg'} />
+        <image height={'28.5%'} href={'/img/va-seal.png'} x={'6.5%'} y={'7.3%'} />
+        <text x={'26.3%'} y={'24.3%'} stroke={'none'} fill={'white'} fontFamily={'Arial, Helvetica, sans-serif'} fontSize={23}>
```
Not sure what font family to use.

rianfowler commented on src/js/vic-v2/components/VeteranIDCard.jsx:
```diff
@@ -1,7 +1,34 @@
+        <image height={'51.8%'} href={'/img/example-photo-2.png'} x={'62.1%'} y={'34.1%'} />
+        <image height={'21.1%'} href={'/img/air-force-coat-of-arms.png'} x={'7.3%'} y={'70%'} />
+        <text x={'7.3%'} y={'55.2%'} stroke={'none'} fill={'white'} fontFamily={'Arial, Helvetica, sans-serif'} fontSize={23}>
+          Firstname Lastname
```
Open to ideas on how to detect breaking this into two lines. Would like to stick to using `<text>` svg elements but that could get tricky to do- maybe measuring text width with a canvas.

cehsu replied:
Could we use number of characters?

rianfowler replied:
We could with a monospaced font. There's a pretty big width difference between 
WWW and 
lll

I know we can render the text to a canvas and measure the width but it might be apples and oranges. There's an package for [word wrapping](https://www.npmjs.com/package/word-wrapper) in the wild or we could just make our own since this use case is very limited. 

We could go further: bump the font size down and measure recursively if the two line name doesn't fit the width constraint or support three line names if someone has a long hyphenated last name. 

Might be helpful to have some actual long last names to test with.

rianfowler commented on src/js/vic-v2/components/VeteranIDCard.jsx:
```diff
@@ -1,7 +1,34 @@
+          Honorable Discharge
+        </text>
+        <text x={'62.1%'} y={'91.9%'} stroke={'none'} fill={'white'} fontFamily={'Arial, Helvetica, sans-serif'} fontSize={12}>
+          ID no: 05P3400000000pz
```
Do we have more examples of what this ID will be?

jbalboni replied:
I _think_ it'll be a case number like `00027471`

rianfowler commented:
The VA logo was not aligned with the name / branch seal. Want to make sure this is intentional. @stacyhlinn 

<img width="266" alt="screen shot 2018-02-08 at 20 07 55" src="https://user-images.githubusercontent.com/4998130/36008201-ceb9514c-0d0b-11e8-8338-a6109fad1795.png">

cehsu commented on src/js/vic-v2/components/VeteranIDCard.jsx:
```diff
@@ -1,7 +1,34 @@
-    <div style={{ paddingTop: '95px', backgroundColor: 'gray', height: '200px', width: '300px' }}></div>
+    <div style={{ marginTop: '50px', backgroundColor: 'grey', height, width: height * WIDTH_HEIGHT_RATIO }}>
+      <svg width={'100%'} height={'100%'}>
+        <image width={'100%'} href={'/img/vic-preview-background.svg'} />
```
Is the number of branches/discharge statuses small enough that it could be worthwhile to pre-generate all possible versions of the card image so that we only need to specify the name, id, and profile photo on the appropriate pre-generated card background?

rianfowler replied:
That would allow the removal 3 text tags and 2 image tags from our code.

cehsu reviewed (commented):
This looks really good so far, pending the questions you mentioned.

rianfowler commented:
Could we show a preview of this on the image cropping page?

jbalboni commented:
@rianfowler I think the designers talked about that and decided the confirmation page was the best place.

jbalboni commented:
@rianfowler Assets can be in the same place as the example photos on the upload/intro page.

Commit:
updated svg

Commit:
add new assets

Commit:
remove test conditions

rianfowler commented on src/js/vic-v2/components/VeteranIDCard.jsx:
```diff
@@ -1,7 +1,106 @@
-  );
+const FONT_FAMILY = 'Arial, Helvetica, sans-serif';
+// sets the aspect ratio of svg
+const VIEW_BOX = '0 0 2048 1284';
```
Measured the mock ups and based the svg view box off of that.

rianfowler commented on src/js/vic-v2/components/VeteranIDCard.jsx:
```diff
@@ -1,7 +1,106 @@
+  <svg
+    fill={'white'}
+    fontFamily={FONT_FAMILY}
+    preserveAspectRatio={'xMidYMid meet'}
```
SVG's container determines its size. Aspect ratio will be maintained.

rianfowler commented on src/js/vic-v2/components/VeteranIDCard.jsx:
```diff
@@ -1,7 +1,106 @@
+// sets the aspect ratio of svg
+const VIEW_BOX = '0 0 2048 1284';
+// height of font top to bottom bounding box in the above viewbox
+const LARGE_FONT_SIZE = 90;
```
Font height is relative to the coordinate system set by the viewbox.

rianfowler commented on src/js/vic-v2/components/VeteranIDCard.jsx:
```diff
@@ -1,7 +1,106 @@
+      .getContext('2d')
+      .measureText(name)
+    // manually determined max width
+      .width > 120;
```
Didn't find a way to do this in the context of the svg's coordinate system so I tested a few widths and got this number. Will test in a few browsers to determine if the value is consistent.

Below represents the max width 1 line name. 

<img width="596" alt="screen shot 2018-02-12 at 20 12 27" src="https://user-images.githubusercontent.com/4998130/36131249-744cb8b0-1036-11e8-8835-2063ed7f86f1.png">
<img width="584" alt="screen shot 2018-02-12 at 20 13 07" src="https://user-images.githubusercontent.com/4998130/36131253-786921ae-1036-11e8-85a6-d57303d06af3.png">

rianfowler commented on src/js/vic-v2/components/VeteranIDCard.jsx:
```diff
@@ -1,7 +1,106 @@
+            {`${veteranFirstName} ${veteranLastName}`}
+          </text>
+        )
+    }
```
This spacing is kind of gross but wanted to make sure it was clear the first condition was returning an array of elements.

rianfowler commented on src/js/vic-v2/components/VeteranIDCard.jsx:
```diff
@@ -1,7 +1,106 @@
+      Honorable Discharge
+    </text>
+
+    <image height={'51.8%'} href={'/img/example-photo-1.png'} x={'62.1%'} y={'34.1%'}/>
```
@jbalboni Not sure how the cropped image is going to be passed to his component.

jbalboni replied:
You will either need to take `photo.serverName` and `photo.serverPath` as props and use the `getImageUrl` function from PhotoField to build the url in here, or you can build the url in ConfirmationPage and pass that in as a prop.

rianfowler replied:
Would I be able to test this locally?

cehsu replied:
Perhaps a mock server could be used similar to how the uploads are tested in the e2e tests.

rianfowler replied:
Good idea- I'll check that out.

rianfowler commented:
@stacyhlinn I've added new screenshots / questions to this PR for review.

jbalboni commented on src/js/vic-v2/components/VeteranIDCard.jsx:
```diff
@@ -1,7 +1,106 @@
+  VASeal: '/img/vic-va-seal.png'
+};
+
+function getIsLongName(name) {
```
Maybe `isLongName`?

jbalboni commented on src/js/vic-v2/components/VeteranIDCard.jsx:
```diff
@@ -1,7 +1,106 @@
+}) => (
+  // svg preserves aspect ratio
+  <svg
+    fill={'white'}
```
Curious why you do this vs `fill="white"` here and elsewhere for string props.

rianfowler replied:
Habit. Will update.

jbalboni commented on src/js/vic-v2/components/VeteranIDCard.jsx:
```diff
@@ -1,7 +1,106 @@
+VeteranIDCard.propTypes = {
+  veteranFirstName: React.PropTypes.string.isRequired,
+  veteranLastName: React.PropTypes.string.isRequired,
+  veteranBranch: React.PropTypes.oneOf(['Air Force', 'Army', 'Coast Guard', 'Marine Corps', 'Navy']),
```
The form data will have the code for these branches, rather than the full description. You'll have to get the labels for those codes in ConfirmationPage or in here (they're in the form config file, but we can pull them into another file), whatever you think is cleaner.

rianfowler commented:
@stacyhlinn Do we want to display middle name if provided? If so, how do we want to name to break if it's long e.g. middle name on first line if it fits, middle name on 2nd line if it doesn't fit, not middle name if it doesn't fit on either.

rianfowler commented:
@stacyhlinn Do we want to show the name in all caps? Do we want to allow uncapitalized names? e.g. rian fowler instead of Rian Fowler.

rianfowler commented:
Do we want to limit what characters veterans can use for their name? e.g. I don't think we stop them from using emoji or other unicode characters.

stacyhlinn commented:
@rianfowler here are what @lalexanderson-dsva and I discussed:

- We'd like to  propose that we keep printing names in one line until there is no more space then line break. so it could be

     Firstname Lastname

     Firstname Middlename Lastname

     Firstname Middlename
     Lastname

     Firstname
     Middlename Lastname

- Please take out "honorable discharge" 
- Add Veteran after the branch name, so for ex: Army Veteran, Coast Guard Veteran...etc 
- Add "TEMPORARY" on the top right corner of the card
- Is it possible to reduce the font size of ID no a bit, so the length doesn't exceed the width of the photo?

![vic-3-03](https://user-images.githubusercontent.com/27507402/36166360-0ec023a8-10a7-11e8-8434-ef3e44947fcd.png)

stacyhlinn commented:
@rianfowler 
- Agreed that we should allow other letters, but no emojis.
- I'd like to enforce first letter cap - Firstname Middlename Lastname

Commit:
wire veteran ID up to confirmation page props and form data

stacyhlinn commented:
@rianfowler 

Sorry I take back what I just wrote in the previous comment. After talking to @lalexanderson-dsva we realized there are scenarios where some names don't always start with a capitalized letter. Let's go ahead and print names in ALL CAPS to get around this problem.

Commit:
update function name

rianfowler commented:
@stacyhlinn Updates:

- Added TEMPORARY to the upper right corner 
- Removed "Honorable Discharge"
- Adjusted branch name position and added "Veteran"
- Names are displayed in all caps 
- Middle name and suffix are included 
- Each name part wraps independently 
- Reduced font size of ID 

<img width="585" alt="screen shot 2018-02-15 at 11 49 48" src="https://user-images.githubusercontent.com/4998130/36274475-74ec2792-124d-11e8-8cf0-b722cc84abda.png">
<img width="587" alt="screen shot 2018-02-15 at 11 51 52" src="https://user-images.githubusercontent.com/4998130/36274476-74fc95d2-124d-11e8-90c9-ba53db2eabbe.png">
<img width="588" alt="screen shot 2018-02-15 at 11 52 04" src="https://user-images.githubusercontent.com/4998130/36274477-750c9d7e-124d-11e8-82d3-6018f9a5a11f.png">
<img width="583" alt="screen shot 2018-02-15 at 11 52 15" src="https://user-images.githubusercontent.com/4998130/36274478-752083d4-124d-11e8-9d9f-42db70e1802c.png">
<img width="581" alt="screen shot 2018-02-15 at 11 52 26" src="https://user-images.githubusercontent.com/4998130/36274479-7532fab4-124d-11e8-8f70-b2a3ddafeb1e.png">
<img width="584" alt="screen shot 2018-02-15 at 11 52 50" src="https://user-images.githubusercontent.com/4998130/36274480-75462f62-124d-11e8-91d7-cb8dfbc45be9.png">

Would like thoughts on how to handle long names. It would be nice if we could match the behavior to the how the printed version will handle it. We have a thirty character limit on each name part which can be pretty long if the name has a lot of wide characters.

<img width="600" alt="screen shot 2018-02-15 at 12 39 47" src="https://user-images.githubusercontent.com/4998130/36274756-5ae6b2d0-124e-11e8-8fb6-139c6a60c535.png">

Commit:
add naming separation logic

Commit:
remove dev test stuff

rianfowler commented on src/js/vic-v2/containers/ConfirmationPage.jsx:
```diff
@@ -41,9 +46,23 @@ class ConfirmationPage extends React.Component {
     const submittedAt = moment(form.submission.submittedAt);
 
+    const veteranFullNameStr =
+      `${firstName.toUpperCase()} ${middleName.toUpperCase()} ${lastName.toUpperCase()} ${suffix.toUpperCase()}` // upper case name
```
Opted for passing the whole name string in rather than the name parts even though the veteranID card breaks them back down into parts. The number of name parts vary (2 minimum up to 3 typically but can be higher with space separated double-barrel last names) so I thought it'd be simpler if the ID component could just assume a space splittable string since the which name part they are isn't relevant to it.

rianfowler commented on src/js/vic-v2/containers/ConfirmationPage.jsx:
```diff
@@ -41,9 +46,23 @@ class ConfirmationPage extends React.Component {
+    const {
+      first: firstName = '',
+      middle: middleName = '',
+      last: lastName = '',
```
Setting some defaults here so I don't have to check if exists when calling `toUpperCase()` later.

Commit:
repair merge conflict

rianfowler commented on src/js/vic-v2/containers/ConfirmationPage.jsx:
```diff
@@ -36,10 +41,24 @@ class ConfirmationPage extends React.Component {
+      middle: middleName = '',
+      last: lastName = '',
+      suffix = ''
+    } = form.data.veteranFullName;
```
Setting some defaults here so I don't have to check if exists when calling `toUpperCase()` later.

rianfowler commented on src/js/vic-v2/containers/ConfirmationPage.jsx:
```diff
@@ -36,10 +41,24 @@ class ConfirmationPage extends React.Component {
     const submittedAt = moment();
 
+    const veteranFullNameStr =
+      `${firstName.toUpperCase()} ${middleName.toUpperCase()} ${lastName.toUpperCase()} ${suffix.toUpperCase()}` // upper case name
```
Opted for passing the whole name string in rather than the name parts even though the veteranID card breaks them back down into parts. The number of name parts vary (2 minimum up to 3 typically but can be higher with space separated double-barrel last names) so I thought it'd be simpler if the ID component could just assume a space splittable string since the which name part they are isn't relevant to it.

rianfowler commented on src/js/vic-v2/containers/ConfirmationPage.jsx:
```diff
@@ -49,7 +68,11 @@ class ConfirmationPage extends React.Component {
+          <VeteranIDCard
+            veteranFullName={veteranFullNameStr}
+            veteranBranchCode={serviceBranch}
+            caseId={this.props.caseId}
```
Assuming this is a prop for now.

jbalboni replied:
It'll be `response.caseId`.

Commit:
fix issue with IE

rianfowler commented on src/sass/vic-v2.scss:
```diff
@@ -99,6 +99,11 @@ input[type=range] {
 
+.id-card-preview {
+  width: 500px;
+  height: 350px;
```
IE 10/11 require a height and width to be set on an svg or its container. Percentages can be used but the image gets scaled to the larger dimension and an svg with a fixed aspect ratio does something like this. 
<img width="756" alt="screen shot 2018-02-16 at 10 26 52" src="https://user-images.githubusercontent.com/4998130/36317895-eeba0e34-1303-11e8-99ae-f74a64842e57.png">

rianfowler commented:
@jbalboni @cehsu This should be ready to merge. Remaining items that'll be handled in future PR:

- handle long names, and
- meeting printing requirements

Commit:
use prop for caseId

rianfowler commented on src/js/vic-v2/containers/ConfirmationPage.jsx:
```diff
@@ -49,7 +68,13 @@ class ConfirmationPage extends React.Component {
+            <VeteranIDCard
+              veteranFullName={veteranFullNameStr}
+              veteranBranchCode={serviceBranch}
+              caseId={this.props.caseId}
```
assuming this will come in as a prop since it's no on the schema.

jbalboni replied:
Think my comment got lost. It'll be `response.caseId`.

Commit:
use response.caseId

jbalboni commented on src/js/vic-v2/components/VeteranIDCard.jsx:
```diff
@@ -1,7 +1,135 @@
+const SMALL_FONT_SIZE = 60;
+const X_SMALL_FONT_SIZE = 55;
+
+const branchCodeLabels = {
```
Can you put this in the vic-v2/labels.js file, and import it from there? You can also update the form config file to pull the labels from there.

jbalboni reviewed (approved):
One minor thing, otherwise LGTM.

Commit:
move branch code labels to labels file

Commit:
fix linting error

Commit:
use a const for max name width
